File: 17_Math_&_Geometry/04_Happy_Number/0202-happy-number.py
"""
Problem: LeetCode 202 - Happy Number

Key Idea:
A number is a happy number if, after repeatedly replacing it with the sum of the square of its digits, the process eventually reaches 1. To determine if a number is a happy number, we can use Floyd's Cycle Detection Algorithm. We use two pointers: a slow pointer that advances one step at a time and a fast pointer that advances two steps at a time. If the fast pointer eventually reaches the slow pointer, we have a cycle, indicating that the number is not a happy number.

Time Complexity:
- The time complexity of the Floyd's Cycle Detection Algorithm is O(log n), where n is the input number.

Space Complexity:
- The space complexity is O(1), as we use a constant amount of space.
"""


class Solution:
    def isHappy(self, n: int) -> bool:
        def get_next(num):
            next_num = 0
            while num > 0:
                num, digit = divmod(num, 10)
                next_num += digit**2
            return next_num

        slow, fast = n, get_next(n)

        while fast != 1 and slow != fast:
            slow = get_next(slow)
            fast = get_next(get_next(fast))

        return fast == 1

File: 17_Math_&_Geometry/04_Happy_Number/0202-happy-number.rs
/*
Problem: LeetCode 202 - Happy Number

Key Idea:
The key idea is to use two pointers, one slow and one fast, to determine if the sequence of numbers ends in a cycle. If the sequence ends in a cycle, it means the number is not a happy number.

Approach:
1. Define a function 'get_next' that calculates the sum of the squares of the digits of a number.
2. Initialize two variables, 'slow' and 'fast', to the given number.
3. Use a loop to repeatedly calculate the next number using 'get_next' for both 'slow' and 'fast'.
4. In each iteration, move 'slow' one step and 'fast' two steps.
5. If 'slow' and 'fast' become equal at any point, break the loop as this indicates a cycle.
6. Check if 'slow' is equal to 1. If it is, return true as the number is a happy number.
7. If the loop completes without finding a cycle and 'slow' is not equal to 1, return false as the number is not a happy number.

Time Complexity:
O(log n), where 'n' is the given number. The time complexity is determined by the number of digits in the input number.

Space Complexity:
O(1), as we are using only a constant amount of extra space.
*/

impl Solution {
    pub fn is_happy(n: i32) -> bool {
        fn get_next(n: i32) -> i32 {
            let mut sum = 0;
            let mut n = n;
            while n > 0 {
                let digit = n % 10;
                sum += digit * digit;
                n /= 10;
            }
            sum
        }

        let mut slow = n;
        let mut fast = n;

        loop {
            slow = get_next(slow);
            fast = get_next(get_next(fast));

            if slow == fast {
                break;
            }
        }

        slow == 1
    }
}

/*
impl Solution {
    pub fn is_happy(n: i32) -> bool {
        let mut n = n;
        while n != 1 && n != 4 {
            let mut sum = 0;
            while n > 0 {
                let d = n % 10;
                sum += d * d;
                n /= 10;
            }
            n = sum;
        }
        n == 1
    }
}
*/
